### PR TITLE
Implement client-side form validation feedback in React forms

### PR DESCRIPTION
**Pull request by flamerohr**
Now that we're on `redux-form` library with `FormBuilder`, lets have some validation added so it doesn't have to go through the server for obvious issues in validation. (e.g. require)

Check out https://github.com/silverstripe/silverstripe-framework/issues/5523 for more details.

Please merge https://github.com/silverstripe/silverstripe-asset-admin/pull/299 as well after this pull request

### CODE REVIEW

**Comment by tractorcow** [comment]
I will review this later today.

**Comment by tractorcow** [comment]
Reviewing now. :)

**Comment by tractorcow** [comment]
Some bugs discovered:

- Fields that have rules but are NOT required somehow become required. E.g. field with `numeric` as a rule but not `required`

image

- Messages for multiple failed rules are not formatted well.

image

**Comment by tractorcow** [comment]
I think the solution to this is to bake into rule lookup some kind of hierarchy:

 - If not required, automatically validate if empty
 - If required AND empty, invalidate, but don't process any other rules.
 - If requried and not empty, process other rules

**Comment by tractorcow** [comment]
Error is traced to schema:
 - form schema ID for the checkboxset field is `Checkboxes`
 - form state ID for the checkbox field is `Form_DetailEditForm_Checkboxes`
 
 So of course it  can't match the IDs to merge state + schema.

**Comment by tractorcow** [comment]
OH! This is because FormField::setSchemaData() acts very bad. If you ever call it, it re-generates the entire schema and stores it. This means that if, later on down the line, the form changes (which would alter the field ID) those changes are no longer reflected.

We can fix this by ensuring that FormField::setSchema() only stores the changed schema, not the full schema.

Before:

```php
public function setSchemaData($schemaData = []) {
	$current = $this->getSchemaData();
	$this->schemaData = array_merge($current, array_intersect_key($schemaData, $current));
	return $this;
}
```

After


```php
public function setSchemaData($schemaData = []) {
	$this->schemaData = array_merge($this->schemaData, array_intersect_key($schemaData, $current));
	return $this;
}
```

Default data will still be merged in getSchemaData(), but not cached.

**Comment by tractorcow** [comment]
Once I fixed that bug, there is still an issue with `required` for multi-select fields.

Checking and unchecking any box treats the field as though it has a value, even though it's empty, and incorrectly allows it to be submitted.


image

**Comment by flamerohr** [comment]
Thanks @tractorcow, I had made the logic simpler "if value is blank and required not defined, don't validate"
and other cases would validate.
And I've fixed the display of each message, they are each on their own line now.

**Comment by tractorcow** [comment]
I've tested and it looks ok.

File: src/ORM/Versioning/ChangeSet.php [hunk 1]
```diff
@@ -399,6 +403,12 @@ public function getCMSFields() {
 		if ($this->isInDB()) {
 			$fields->addFieldToTab('Root.Main', ReadonlyField::create('State', $this->fieldLabel('State')));
 		}
+		$fields->addFieldToTab('Root.Main', $field = new TextField('OptionalField'));
```
Comment (tractorcow) [inline src/ORM/Versioning/ChangeSet.php]: Temp code here needs deleting before merge.

Comment (flamerohr) [inline src/ORM/Versioning/ChangeSet.php]: Done :D

**Pull request merged**